File: leetcode/python/97_interleaving_string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False
        
        dp = [False] * (len(s2) + 1)
        dp[0] = True
        for j in range(1, len(s2) + 1):
            if s2[j-1] == s3[j-1]:
                dp[j] = True
            else:
                break
        
        for i in range(1, len(s1) + 1):
            dp[0] = dp[0] and (s1[i-1] == s3[i-1])
            for j in range(1, len(s2) + 1):
                dp[j] = (dp[j-1] and s2[j-1] == s3[i+j-1]) or (dp[j] and s1[i-1] == s3[i+j-1])
        return dp[-1]

s = Solution()
print(s.isInterleave(s1 = "aabcc", s2 = "dbbca", s3 = "aadbbcbcac"))
print(s.isInterleave(s1 = "aabcc", s2 = "dbbca", s3 = "aadbbbaccc"))

File: leetcode/rust/src/97_interleaving_string.rs
impl Solution {
    pub fn is_interleave(s1: String, s2: String, s3: String) -> bool {
        let m = s1.len();
        let n = s2.len();
        let s1 = s1.chars().collect::<Vec<char>>();
        let s2 = s2.chars().collect::<Vec<char>>();
        let s3 = s3.chars().collect::<Vec<char>>();

        let mut dp = vec![vec![false; n + 1]; m + 1];

        if m + n != s3.len() {
            return false;
        }

        dp[0][0] = true;
        for j in 1..=n {
            dp[0][j] = dp[0][j - 1] && s2[j - 1] == s3[j - 1];
        }
        for i in 1..=m {
            dp[i][0] = dp[i - 1][0] && s1[i - 1] == s3[i - 1];
        }

        for i in 1..=m {
            for j in 1..=n {
                if dp[i-1][j] && s1[i-1] == s3[i+j-1] {
                    dp[i][j] = true;
                }
                else if dp[i][j-1] && s2[j-1] == s3[i+j-1] {
                    dp[i][j] = true;
                }
            }
        }

        return dp[m][n];
    }
}